lib/common/url.c

```diff
@@ -40,97 +40,88 @@ static int decode_hex(int ch)
+/* Perform path normalization and URL decoding in one pass.
+ * @norm_indexes is used to store a mapping from the characters in @path
+ * to the ones in the returned iovec. */
+static h2o_iovec_t rebuild_path(h2o_mem_pool_t *pool, const char *path, size_t len, size_t *query_at, size_t **norm_indexes)
 {
+    const char *s = path;
     size_t i;
     h2o_iovec_t ret;
     char *dst;
+    const char *last_slash;
+    size_t *nindexes;
 
-    dst = ret.base = h2o_mem_alloc_pool(pool, len + 1);
-
-    /* decode %xx */
-    for (i = 0; i + 3 <= len;) {
-        int hi, lo;
-        if (s[i] == '%' && (hi = decode_hex(s[i + 1])) != -1 && (lo = decode_hex(s[i + 2])) != -1) {
-            *dst++ = (hi << 4) | lo;
-            i += 3;
+    { /* locate '?', and set len to the end of input path */
+        const char *q = memchr(path, '?', len);
+        if (q != NULL) {
+            len = *query_at = q - path;
         } else {
-            *dst++ = s[i++];
+            *query_at = SIZE_MAX;
         }
     }
-    while (i != len)
-        *dst++ = s[i++];
-    *dst = '\0';
-
-    ret.len = dst - ret.base;
-    return ret;
-}
 
-static h2o_iovec_t rewrite_special_paths(h2o_mem_pool_t *pool, const char *src, size_t len)
-{
-    h2o_iovec_t ret;
-    const char *src_end = src + len;
-    char *dst;
+    dst = ret.base = h2o_mem_alloc_pool(pool, len);
+    *norm_indexes = nindexes = h2o_mem_alloc_pool(pool, len * sizeof(*norm_indexes[0]));
+    *dst++ = '/';
 
-    if (src == src_end)
-        return h2o_iovec_init("/", 1);
+    if (*s == '/') {
+        s++;
+        len--;
+    }
+    *nindexes++ = s - path;
+    last_slash = ret.base;
 
-    dst = ret.base = h2o_mem_alloc_pool(pool, len + 1);
+    /* decode %xx */
+    for (i = 0; i < len;) {
+        int hi, lo;
+        char decoded;
 
-    /* assertion hereafter: ret.base[0] is always '/' */
-    *dst++ = '/';
-    if (src[0] == '/')
-        ++src;
-
-    /* split by '/' and handle, at the same time combining repeating '/'s to one */
-    while (1) {
-        const char *part_start = src;
-        for (; src != src_end; ++src)
-            if (*src == '/')
-                break;
-        size_t part_len = src - part_start;
-        if (part_len == 1 && part_start[0] == '.')
-            continue;
-        if (part_len == 2 && memcmp(part_start, "..", 2) == 0) {
-            /* if we can go back one level, do it */
-            if (ret.base != dst - 1) {
-                for (--dst; dst[-1] != '/'; --dst)
-                    ;
+        if (s[i] == '%' && (i + 2 < len) && (hi = decode_hex(s[i + 1])) != -1 && (lo = decode_hex(s[i + 2])) != -1) {
+            decoded = (hi << 4) | lo;
+            i += 3;
+        } else {
+            decoded = s[i++];
+        }
+        if (decoded == '/') {
+            size_t part_size = dst - last_slash;
+            if (part_size == 2 && dst[-1] == '.') {
+                dst-=1;
+                nindexes-=1;
+                last_slash = dst - 1;
```

Comment:
Do you think it would be a good idea to define this logic erasing `./` as a macro and use it at the tail of the function for handling `.` at the end of the function?

I believe same thing could also be done for between the handling of `../` and `..` at the end of the function.